wireguard/wireguard.go

```diff
@@ -1256,92 +1306,86 @@ func (w *Wireguard) ensureLinkAddressV4(netlinkClient netlinkshim.Netlink) error
+		w.localCIDRs.Iter(func(itemCIDR interface{}) error {
+			cidr := itemCIDR.(ip.CIDR)
+			newFiltered.Add(cidr)
+			if oldFiltered.Contains(cidr) {
+				oldFiltered.Discard(cidr)
+			} else {
+				w.routerule.SetRule(w.createRouteRule(cidr))
 			}
-
-			// Rule does not match expected, delete it.
-			if err := netlinkClient.RuleDel(&rule); err != nil {
-				w.logCxt.WithError(err).Error("Unable to delete wireguard routing rule")
-				return err
+			return nil
+		})
+		w.localIPs.Iter(func(itemAddr interface{}) error {
+			addr := itemAddr.(ip.Addr)
+			overlaps := false
+			w.localCIDRs.Iter(func(itemCIDR interface{}) error {
+				cidr := itemCIDR.(ip.CIDR).ToIPNet()
+				if cidr.Contains(addr.AsNetIP()) {
+					overlaps = true
+					return set.StopIteration
+				}
+				return nil
+			})
+			if !overlaps {
+				ipAsCidr := addr.AsCIDR()
+				newFiltered.Add(ipAsCidr)
+				if oldFiltered.Contains(ipAsCidr) {
+					oldFiltered.Discard(ipAsCidr)
+				} else {
+					w.routerule.SetRule(w.createRouteRule(ipAsCidr))
+				}
 			}
-		}
-	}
-
-	if found {
-		return nil
-	}
+			return nil
+		})
+		oldFiltered.Iter(func(itemCIDR interface{}) error {
+			cidr := itemCIDR.(ip.CIDR)
+			w.routerule.RemoveRule(w.createRouteRule(cidr))
+			return nil
+		})
 
-	// Add the missing rule.
-	if err := netlinkClient.RuleAdd(newrule); err != nil {
-		w.logCxt.WithError(err).Error("Unable to create wireguard routing rule")
-		return err
-	} else {
-		w.logCxt.Debugf("Added rule: %#v", newrule)
+		w.localCIDRsFiltered = newFiltered
+		w.localCIDRsUpdated = false
 	}
 
-	return nil
+	// Apply the routing rule updates.
+	return w.routerule.Apply()
 }
 
-// ensureNoRouteRule ensures the ip rule to jump to the wireguard table is configured. Since only the wireguard
-// module should be using the wireguard table (important to prevent routing loops), this method deletes any rules that
-// jump to that table and do not match the rule spec expected by this module, and will create the appropriate rule
-// if missing.
-func (w *Wireguard) ensureNoRouteRule(netlinkClient netlinkshim.Netlink) error {
-	// Get the programmed rules.
-	rules, err := netlinkClient.RuleList(netlink.FAMILY_V4)
-	if err != nil {
-		return err
-	}
-
-	for _, rule := range rules {
-		if rule.Table == w.config.RoutingTableIndex {
-			w.logCxt.Debugf("Found rule to table %d", w.config.RoutingTableIndex)
-
-			// Rule does not match expected, delete it.
-			if err := netlinkClient.RuleDel(&rule); netlinkshim.IsNotExist(err) {
-				w.logCxt.Debug("Wireguard routing rule already deleted")
-			} else if err != nil {
-				w.logCxt.WithError(err).Error("Unable to delete wireguard routing rule")
-				return err
-			}
-		}
-	}
-
-	return nil
+// createRouteRule creates a routing rule to route a local source CIDR to the wireguard table (if wireguard firewall
+// mark is not set).
+func (w *Wireguard) createRouteRule(cidr ip.CIDR) *routerule.Rule {
+	rule := routerule.NewRule(ipVersion, w.config.RoutingRulePriority).
```

Comment:
Is it OK to have multiple rules with the same priority?

Reply:
Yup